File: go-backend/internal/handlers/tasks.go
package handlers

import (
	"net/http"
	"strconv"

	"go-backend/internal/database"
	"go-backend/internal/middleware"
	"go-backend/internal/models"

	"github.com/gin-gonic/gin"
)

// ListTasks はユーザーのタスク一覧を返す
func ListTasks(c *gin.Context) {
	userID := middleware.GetCurrentUserID(c)
	if userID == 0 {
		c.JSON(http.StatusUnauthorized, gin.H{
			"detail": "Authentication credentials were not provided.",
		})
		return
	}

	db := database.GetDB()

	var tasks []models.Task
	if err := db.Where("user_id = ?", userID).Order("created_at DESC").Find(&tasks).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"detail": "Failed to retrieve tasks.",
		})
		return
	}

	// レスポンス形式に変換
	response := make([]models.TaskResponse, len(tasks))
	for i, task := range tasks {
		response[i] = task.ToResponse()
	}

	c.JSON(http.StatusOK, response)
}

// CreateTask は新しいタスクを作成
func CreateTask(c *gin.Context) {
	userID := middleware.GetCurrentUserID(c)
	if userID == 0 {
		c.JSON(http.StatusUnauthorized, gin.H{
			"detail": "Authentication credentials were not provided.",
		})
		return
	}

	var req models.CreateTaskRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"detail": "Invalid request data.",
			"errors": err.Error(),
		})
		return
	}

	db := database.GetDB()

	task := models.Task{
		Title:       req.Title,
		Description: req.Description,
		Completed:   req.Completed,
		UserID:      userID,
	}

	if err := db.Create(&task).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"detail": "Failed to create task.",
		})
		return
	}

	c.JSON(http.StatusCreated, task.ToResponse())
}

// GetTask は特定のタスクを取得
func GetTask(c *gin.Context) {
	userID := middleware.GetCurrentUserID(c)
	if userID == 0 {
		c.JSON(http.StatusUnauthorized, gin.H{
			"detail": "Authentication credentials were not provided.",
		})
		return
	}

	taskID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"detail": "Invalid task ID.",
		})
		return
	}

	db := database.GetDB()

	var task models.Task
	if err := db.Where("id = ? AND user_id = ?", taskID, userID).First(&task).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"detail": "Task not found.",
		})
		return
	}

	c.JSON(http.StatusOK, task.ToResponse())
}

// UpdateTask はタスクを更新（部分更新対応）
func UpdateTask(c *gin.Context) {
	userID := middleware.GetCurrentUserID(c)
	if userID == 0 {
		c.JSON(http.StatusUnauthorized, gin.H{
			"detail": "Authentication credentials were not provided.",
		})
		return
	}

	taskID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"detail": "Invalid task ID.",
		})
		return
	}

	db := database.GetDB()

	// タスク取得
	var task models.Task
	if err := db.Where("id = ? AND user_id = ?", taskID, userID).First(&task).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"detail": "Task not found.",
		})
		return
	}

	var req models.UpdateTaskRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"detail": "Invalid request data.",
			"errors": err.Error(),
		})
		return
	}

	// 部分更新
	updates := make(map[string]interface{})
	if req.Title != nil {
		updates["title"] = *req.Title
	}
	if req.Description != nil {
		updates["description"] = *req.Description
	}
	if req.Completed != nil {
		updates["completed"] = *req.Completed
	}

	if len(updates) > 0 {
		if err := db.Model(&task).Updates(updates).Error; err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"detail": "Failed to update task.",
			})
			return
		}
	}

	// 更新後のタスクを取得
	db.First(&task, taskID)

	c.JSON(http.StatusOK, task.ToResponse())
}

// DeleteTask はタスクを削除
func DeleteTask(c *gin.Context) {
	userID := middleware.GetCurrentUserID(c)
	if userID == 0 {
		c.JSON(http.StatusUnauthorized, gin.H{
			"detail": "Authentication credentials were not provided.",
		})
		return
	}

	taskID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"detail": "Invalid task ID.",
		})
		return
	}

	db := database.GetDB()

	// タスクの存在確認
	var task models.Task
	if err := db.Where("id = ? AND user_id = ?", taskID, userID).First(&task).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"detail": "Task not found.",
		})
		return
	}

	// 削除（物理削除）
	if err := db.Unscoped().Delete(&task).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"detail": "Failed to delete task.",
		})
		return
	}

	c.Status(http.StatusNoContent)
}

File: rust-backend/src/handlers/tasks.rs
use actix_web::{web, HttpResponse};
use chrono::Utc;
use serde_json::json;
use validator::Validate;

use crate::db::DbPool;
use crate::models::{CreateTaskRequest, Task, TaskResponse, UpdateTaskRequest, User};

/// GET /api/tasks/
pub async fn list_tasks(pool: web::Data<DbPool>, user: User) -> HttpResponse {
    let tasks: Vec<Task> =
        sqlx::query_as("SELECT * FROM tasks WHERE user_id = ? ORDER BY created_at DESC")
            .bind(user.id)
            .fetch_all(pool.get_ref())
            .await
            .unwrap_or_default();

    let response: Vec<TaskResponse> = tasks.into_iter().map(|t| t.into()).collect();
    HttpResponse::Ok().json(response)
}

/// POST /api/tasks/
pub async fn create_task(
    pool: web::Data<DbPool>,
    user: User,
    body: web::Json<CreateTaskRequest>,
) -> HttpResponse {
    if let Err(e) = body.validate() {
        return HttpResponse::BadRequest().json(json!({
            "detail": "Invalid request data.",
            "errors": e.to_string()
        }));
    }

    let now = Utc::now();
    let description = body.description.clone().unwrap_or_default();
    let completed = body.completed.unwrap_or(false);

    let result = sqlx::query(
        "INSERT INTO tasks (title, description, completed, created_at, updated_at, user_id) VALUES (?, ?, ?, ?, ?, ?)",
    )
    .bind(&body.title)
    .bind(&description)
    .bind(completed)
    .bind(now)
    .bind(now)
    .bind(user.id)
    .execute(pool.get_ref())
    .await;

    match result {
        Ok(r) => {
            let task_id = r.last_insert_rowid();
            HttpResponse::Created().json(TaskResponse {
                id: task_id,
                title: body.title.clone(),
                description,
                completed,
                created_at: now,
                updated_at: now,
                user: user.id,
            })
        }
        Err(_) => HttpResponse::InternalServerError().json(json!({
            "detail": "Failed to create task."
        })),
    }
}

/// GET /api/tasks/{id}/
pub async fn get_task(
    pool: web::Data<DbPool>,
    user: User,
    path: web::Path<i64>,
) -> HttpResponse {
    let task_id = path.into_inner();

    let task: Option<Task> =
        sqlx::query_as("SELECT * FROM tasks WHERE id = ? AND user_id = ?")
            .bind(task_id)
            .bind(user.id)
            .fetch_optional(pool.get_ref())
            .await
            .unwrap_or(None);

    match task {
        Some(t) => HttpResponse::Ok().json(TaskResponse::from(t)),
        None => HttpResponse::NotFound().json(json!({
            "detail": "Task not found."
        })),
    }
}

/// PATCH /api/tasks/{id}/
pub async fn update_task(
    pool: web::Data<DbPool>,
    user: User,
    path: web::Path<i64>,
    body: web::Json<UpdateTaskRequest>,
) -> HttpResponse {
    let task_id = path.into_inner();

    // Check task exists and belongs to user
    let task: Option<Task> =
        sqlx::query_as("SELECT * FROM tasks WHERE id = ? AND user_id = ?")
            .bind(task_id)
            .bind(user.id)
            .fetch_optional(pool.get_ref())
            .await
            .unwrap_or(None);

    let task = match task {
        Some(t) => t,
        None => {
            return HttpResponse::NotFound().json(json!({
                "detail": "Task not found."
            }));
        }
    };

    // Build update
    let title = body.title.clone().unwrap_or(task.title);
    let description = body.description.clone().unwrap_or(task.description);
    let completed = body.completed.unwrap_or(task.completed);
    let now = Utc::now();

    let result = sqlx::query(
        "UPDATE tasks SET title = ?, description = ?, completed = ?, updated_at = ? WHERE id = ?",
    )
    .bind(&title)
    .bind(&description)
    .bind(completed)
    .bind(now)
    .bind(task_id)
    .execute(pool.get_ref())
    .await;

    match result {
        Ok(_) => HttpResponse::Ok().json(TaskResponse {
            id: task_id,
            title,
            description,
            completed,
            created_at: task.created_at,
            updated_at: now,
            user: user.id,
        }),
        Err(_) => HttpResponse::InternalServerError().json(json!({
            "detail": "Failed to update task."
        })),
    }
}

/// DELETE /api/tasks/{id}/
pub async fn delete_task(
    pool: web::Data<DbPool>,
    user: User,
    path: web::Path<i64>,
) -> HttpResponse {
    let task_id = path.into_inner();

    // Check task exists and belongs to user
    let task: Option<Task> =
        sqlx::query_as("SELECT * FROM tasks WHERE id = ? AND user_id = ?")
            .bind(task_id)
            .bind(user.id)
            .fetch_optional(pool.get_ref())
            .await
            .unwrap_or(None);

    if task.is_none() {
        return HttpResponse::NotFound().json(json!({
            "detail": "Task not found."
        }));
    }

    let result = sqlx::query("DELETE FROM tasks WHERE id = ?")
        .bind(task_id)
        .execute(pool.get_ref())
        .await;

    match result {
        Ok(_) => HttpResponse::NoContent().finish(),
        Err(_) => HttpResponse::InternalServerError().json(json!({
            "detail": "Failed to delete task."
        })),
    }
}
